app/models/name.rb

```diff
@@ -408,33 +457,196 @@ class Name < AbstractModel
+            or(where(NameDescription[:notes].matches("%#{text}%"))).
+            or(where(NameDescription[:refs].matches("%#{text}%")))
+        }
+  scope :with_description_in_project,
+        lambda { |project|
+          joins(descriptions: :project).
+            merge(NameDescription.where(project: project))
+        }
+  scope :with_description_created_by,
+        lambda { |user|
+          joins(:descriptions).
+            merge(NameDescription.where(user: user))
+        }
+  scope :with_description_reviewed_by,
+        lambda { |user|
+          joins(:descriptions).
+            merge(NameDescription.where(reviewer: user))
+        }
+  scope :with_description_of_type,
+        lambda { |source|
+          # Check that it's a valid source type (string enum value)
+          return Name.none if Description.all_source_types.exclude?(source)
+
+          joins(:descriptions).
+            merge(NameDescription.where(source_type: source))
+        }
 
   ### Module Name::Spelling
   scope :with_correct_spelling, -> { where(correct_spelling_id: nil) }
-
-  # For a glitch discovered in the wild:
+  scope :with_incorrect_spelling, -> { where.not(correct_spelling_id: nil) }
   scope :with_self_referential_misspelling, lambda {
     where(Name[:correct_spelling_id].eq(Name[:id]))
   }
+  scope :with_synonyms, -> { where.not(synonym_id: nil) }
+  scope :without_synonyms, -> { where(synonym_id: nil) }
+  scope :ok_for_export, -> { where(ok_for_export: true) }
 
   ### Module Name::Taxonomy
-  scope :with_classification_like,
-        # Use multi-line lambda literal because fixtures blow up with "lambda":
-        # NoMethodError: undefined method `ranks'
-        #   test/fixtures/names.yml:28:in `get_binding'
-        ->(rank, text_name) { # rubocop:disable Style/Lambda
+  scope :with_rank,
+        ->(rank) { where(rank: Name.ranks[rank]) if rank }
+  scope :with_rank_below,
+        ->(rank) { where(Name[:rank] < Name.ranks[rank]) if rank }
+  scope :with_rank_and_name_in_classification,
+        lambda { |rank, text_name|
           where(Name[:classification].matches("%#{rank}: _#{text_name}_%"))
         }
+  scope :with_rank_at_or_below_genus,
+        lambda {
+          where((Name[:rank] <= Name.ranks[:Genus]).
+                or(Name[:rank] == Name.ranks[:Group]))
+        }
+  scope :with_rank_above_genus,
+        lambda {
+          where(Name[:rank] > Name.ranks[:Genus]).
+            where(Name[:rank] != Name.ranks[:Group])
+        }
+  scope :subtaxa_of_genus_or_below,
+        lambda { |text_name|
+          text_name_followed_by_group =
+            /#{text_name} (#{Name::Parse::GROUP_ABBR.source.delete(" ")})/
+
+          # Note small diff w :text_name_includes scope
+          where(Name[:text_name].matches("#{text_name} %")).
+            # "<name> group" is not a subtaxon of "<name>"
+            where.not(Name[:text_name] =~ text_name_followed_by_group)
+        }
```

Comment:
Questions about `scope :subtaxa_of_genus_or_below`:
### Is `<name> group` is a subtaxon of `<name>`?
Logically`<name> group` is a superset, not a subtaxon, of `<name>`. 
But maybe MO should treat it as a subtaxon

### Is there a reasonable DRY, but more performant way of excluding `<name> group` from the results?
This question assumes that `<name> group` is not subtaxon of `<name>`.
- By DRY, I mean that the query depends on `Name::Parse::GROUP_ABBR`.
- I'm worried about the performance of 
  - constructing a regex each time this scope is called;
  - the performance of SQL that uses regular expression mapping

Note: I had started with the following hard-coded string before turning to the regex stuff:
```ruby
where.not(text_name: "#{text_name} group").
where.not(text_name: "#{text_name} gr.").
where.not(text_name: "#{text_name} gr").
where.not(text_name: "#{text_name} gp.").
where.not(text_name: "#{text_name} gp").
where.not(text_name: "#{text_name} clade").
where.not(text_name: "#{text_name} complex")
```

Reply:
re: Is <name> group a subset of <name> -- Definitely not.  I would say the majority if not all cases of <name> group, <name> is a subtaxon of <name> group.  The general use case (at least how I always use it personally, and the only way I've ever seen it used in lichenology) is when a set of species is similar and hard to distinguish -- sometimes related and forming a monophyletic clade, sometimes only superficially similar and therefore polyphyletic.  Someone will choose an exemplar and refer to the entire set of similar species as the "So andso group".  Lecanora dispersa group, for example, turned out to be "good" (monophyletic) and got named Myriolecis eventually.  Lecidea atrobrunnea group, on the other hand, is a highly diverse set of superficially similar species in multiple families.  But in both of these cases Lecanora dispersa is a subset of the L. dispersa group, and L. atrobrunnea is a subset of the L. atrobrunnea group.

*But*... We have never codified anywhere which species belong in a group.  In fact, people rarely even bother to explicitly deffine what the group means even in comments under the group name itself!  Therefore it is impossible to get the subtaxa scopes to work with groups.

I would just ignore the problem until it goes away, personally.

(Which in practice probably means that the group will show up as a subtaxon of the name, unless we explicitly exclude groups from the results as a matter of course.  The latter may be the less confusing, safer course, but I leave it to you.  It's not that important to me.  Groups are wishy-washy, "non-real" names in the best of circumstances.  I'm not concerned if their behavior is erratic, ill-defined, or non-intuitive.)

Also, by the way, there should be no reason to use a regular expression on text_name to determine if a taxon is a group -- just check the rank.  Also, by the way, the name parser should automatically normalize the spelling.  I think it allows only "group", "clade" and maybe "complex"(?) and "agg."(?) (for aggregate).

Reply:
Thanks for all the comments.

When I looked at the draft of this scope, I thought that treating `X group` as a subtaxon of `X` was a subtle bug, so I wrote tests and code to avoid that. I'll think about whether to retain that stuff.

If I do retain it,  I can't exclude all groups from the results, and therefore can't just check the rank.
Consider, for instance subtaxa_of_genus("Boletus"). That should include groups in ranks below genus, e.g., Boletus edulis group. The only group that should be excluded is "Boletus group". 
So I think the logic necessarily requires looking the text_names. 

(And if I do retain it, I'll check out your suggestion about normalized spelling.)

Reply:
Responding to Joe -- re: "I can't exclude all groups from the results, and therefore can't just check the rank.
Consider, for instance subtaxa_of_genus("Boletus"). That should include groups in ranks below genus, e.g., Boletus edulis group. The only group that should be excluded is "Boletus group"."

I would just like to re-emphasize that it is *impossible* to do this right.  I agree with you that we can twiddle with a few edge cases and make it slightly "less wrong".  But it's worth asking whether the effort and presumed loss in performance is worth it for such minimal gains.

Queries including subtaxa are actually common.  At least I have used them numerous times.  This is not an academic question.